brainiak/reprsimil/brsa.py

```diff
@@ -0,0 +1,1118 @@
+            design = np.concatenate((design,
+                                     np.ones([design.shape[0], 1])), axis=1)
+        assert self.rank is None or self.rank <= design.shape[1],\
+            'Your design matrix has fewer columns than the rank you set'
+
+        # check scan_onsets validity
+        assert scan_onsets is None or\
+            (np.max(scan_onsets) <= X.shape[0] and np.min(scan_onsets) >= 0),\
+            'Some scan onsets provided are out of the range of time points.'
+
+        # check the size of coords and inten
+        if self.GP_space:
+            logger.debug('Fitting with Gaussian Process prior on log(SNR)')
+            assert coords is not None and coords.shape[0] == X.shape[1],\
+                'Spatial smoothness was requested by setting GP_space. '\
+                'But the voxel number of coords does not match that of '\
+                'data X, or voxel coordinates are not provided. '\
+                'Please make sure that coords is in the shape of '\
+                '[ n_voxel x 3].'
+            assert coords.ndim <= 2,\
+                'The coordinate matrix should not have '\
+                'more than 2 dimensions'
+            if coords.ndim == 1:
+                coords = np.expand_dims(coords, 1)
+            if self.GP_inten:
+                assert inten is not None and inten.shape[0] == X.shape[1],\
+                    'The voxel number of intensity does not '\
+                    'match that of data X, or intensity not provided.'
+                assert np.var(inten) > 0,\
+                    'All voxels have the same intensity.'
+                if inten_weight is None:
+                    inten_weight = self.inten_weight
+                assert inten_weight >= 0 and inten_weight <= 1,\
+                    'The weight of intensity kernel is '\
+                    'out of the range of (0,1).'\
+                    'I suggest using a small value, such as 0.1.'
+            else:
+                inten_weight = 0.0
+
+        # Run Bayesian RSA
+        if not self.GP_space:
+            self.U_, self.L_, self.nSNR_, self.sigma_, self.rho_ = \
+                self._fit_RSA_UV(design, X, scan_onsets=scan_onsets)
+        elif not self.GP_inten or inten_weight == 0:
+            self.U_, self.L_, self.nSNR_, self.sigma_, self.rho_,\
+                self.lGPspace_, self.bGP_ = self._fit_RSA_UV(
+                    design, X, scan_onsets=scan_onsets, coords=coords)
+        else:
+            self.U_, self.L_, self.nSNR_, self.sigma_, self.rho_, \
+                self.lGPspace_, self.bGP_, self.lGPinten_ = \
+                self._fit_RSA_UV(design, X, scan_onsets=scan_onsets,
+                                 coords=coords, inten=inten,
+                                 inten_weight=inten_weight)
+
+        if self.pad_DC:
+            self.U_ = self.U_[:-1, :-1]
+            self.L_ = self.L_[:-1, :self.rank]
+        sqrt_diag = np.diag(self.U_)**0.5
```

Comment:
Maybe more elegant: 
```
inv_sd = 1/np.sqrt(np.diag(self.U_))
self.C_ = self.U_ * np.outer(inv_sd, inv_sd)
```
(or even better yet, define `__cov2cor` elsewhere and call it here)